source/common/network/io_socket_handle_impl.cc

```diff
@@ -466,7 +475,28 @@ IoHandlePtr IoSocketHandleImpl::accept(struct sockaddr* addr, socklen_t* addrlen
 }
 
 Api::SysCallIntResult IoSocketHandleImpl::connect(Address::InstanceConstSharedPtr address) {
-  return Api::OsSysCallsSingleton::get().connect(fd_, address->sockAddr(), address->sockAddrLen());
+  auto sockaddr_to_use = address->sockAddr();
+  auto sockaddr_len_to_use = address->sockAddrLen();
+#ifdef __ANDROID_API__
+  sockaddr_in6 sin6;
+  if (sockaddr_to_use->sa_family == AF_INET && alwaysUseV6OnAndroid()) {
+    const sockaddr_in& sin4 = reinterpret_cast<const sockaddr_in&>(*sockaddr_to_use);
+
+    // Android always uses IPv6 dual stack. Convert IPv4 to the IPv6 mapped address when
+    // connecting.
+    memset(&sin6, 0, sizeof(sin6));
+    sin6.sin6_family = AF_INET6;
+    sin6.sin6_port = sin4.sin_port;
+    sin6.sin6_addr.s6_addr32[2] = htonl(0xffff);
+    sin6.sin6_addr.s6_addr32[3] = sin4.sin_addr.s_addr;
+    ASSERT(IN6_IS_ADDR_V4MAPPED(&sin6.sin6_addr));
+
+    sockaddr_to_use = reinterpret_cast<sockaddr*>(&sin6);
+    sockaddr_len_to_use = sizeof(sin6);
```

Comment:
I think the addresses we get from sockets are now going to be of the style 
ffff:1.2.3.4 which I believe is controlled via config v6only.  Should we also default sanitize getting addresses so we strip the ffff (v6only or forceNormalizeV6ForAndroid) ?

Reply:
I need to look and see if we ever actually pull the the peer address. I'm not sure that we do but let me look further.

Reply:
So yeah we do have `Address::InstanceConstSharedPtr IoSocketHandleImpl::peerAddress()`, but I don't think we ever use this in the client context. Let me add a comment/TODO there in case this becomes an issue. I would rather not make this change more complicated than it needs to be since I don't think I can test it (see other comment).